# Preprocessing

In [2]:
import pandas as pd
import numpy as np

In [3]:
from functions import *

### Import the datasets

In [4]:
# Open the raw datasets
# Heavy Metal
hm_df = pd.read_excel('data/AllHMFiles.xlsx', engine='openpyxl').fillna(0.0)

# Pesticides
pesticides_df = pd.read_excel('data/AllPesticidesFiles.xlsx', engine='openpyxl').fillna(0.0)
classe_pesticides_df = pd.read_excel('data/pesticides.xlsx', engine='openpyxl')

# Surfaces
surfaces_df = pd.read_excel('data/absSurfs.xlsx', engine='openpyxl')

# Distances
distances_df = pd.read_excel('data/distsOneSheet.xlsx', engine='openpyxl')

In [5]:
hm_df.head(5)

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  
0     0.032    40     0.0  
1     0.029    33     0.0  
2     0.038    37     0.0  
3     0.031    37     0.0  
4     0.052    35     0.0

In [6]:
pesticides_df.head(5)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
0            B0017  2017/P1     0.0     0.0000          0.00           0.0   
1            B0019  2017/P1     0.0     0.0188          0.00           0.0   
2            B0021  2017/P1     0.0     0.0711          0.00           0.0   
3            B0023  2017/P1     0.0     0.0294          0.00           0.0   
4            B0037  2017/P1     0.0     0.0000          0.01           0.0   

   Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  Nicosulfuron  \
0           0.0           0.0          0.0        0.0  ...           0.0   
1           0.0           0.0          0.0        0.0  ...           0.0   
2           0.0           0.0          0.0        0.0  ...           0.0   
3           0.0           0.0          0.0        0.0  ...           0.0   
4           0.0           0.0          0.0        0.0  ...           0.0   

   Piperonylbutoxide  Prosulfuron  Simazine  Terbuthylazine  Dieldrin  \
0                0.0          0.0       0.0             0.0       0.0   
1                0.0          0.0       0.0             0.0       0.0   
2                0.0          0.0       0.0             0.0       0.0   
3                0.0          0.0       0.0             0.0       0.0   
4                0.0          0.0       0.0             0.0       0.0   

   Fenamidone  Trifluralin  Cycloate  Pentachloroanisol  
0         0.0          0.0       0.0                0.0  
1         0.0          0.0       0.0                0.0  
2         0.0          0.0       0.0                0.0  
3         0.0          0.0       0.0                0.0  
4         0.0          0.0       0.0                0.0  

[5 rows x 117 columns]

In [7]:
classe_pesticides_df.head(5)

importName           name    statusEN   LMR  \
0      Aclonifen      aclonifen  authorized  0.05   
1         Aldrin         aldrin  prohibited  0.01   
2   Ametoctradin   ametoctradin  authorized  0.05   
3  Anthraquinone  anthraquinone  prohibited  0.02   
4   Azoxystrobin   azoxystrobin  authorized  0.05   

                          familyEN          typeEN  
0                   diphenyl ether       herbicide  
1                   organochlorine     insecticide  
2               triazolopyrimidine       fungicide  
3  polycyclic aromatic hydrocarbon  bird repellant  
4                      strobilurin       fungicide

In [8]:
surfaces_df.head(5)

Site  111           112           121            122  123  124  131  132  \
0  F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1  F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2  F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3  B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4  F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   

   133  ...  411  412  421  422  423  511            512  521  522  523  
0  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4  0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  

[5 rows x 45 columns]

In [9]:
distances_df.head(5)

Site     polyID  classCLC         dist
0  F0069   FR-88090       231  1183.022362
1  F0069  FR-131773       242   289.926324
2  F0069   FR-46167       211   221.344153
3  F0069   FR-11047       112  1038.383069
4  F0069  FR-195343       311   724.045683

### Heavy Metal

In [10]:
hm_df.head()

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  
0     0.032    40     0.0  
1     0.029    33     0.0  
2     0.038    37     0.0  
3     0.031    37     0.0  
4     0.052    35     0.0

In [54]:
MRL_HM_dict = {
    'Lead' : [1.5],
    'Cadmium' : [1],
    'Arsenic' : [0.3],
    'Mercury' : [1],
    'Copper' : [30]
}

MRL_HM = pd.DataFrame.from_dict(MRL_HM_dict)
MRL_HM

Lead  Cadmium  Arsenic  Mercury  Copper
0   1.5        1      0.3        1      30

In [69]:
hm_df['mrl_lead'] = np.where(hm_df['Lead'] > 1.5, 1, 0)
hm_df['mrl_cadmium'] = np.where(hm_df['Cadmium'] > 1, 1, 0)
hm_df['mrl_arsenic'] = np.where(hm_df['Arsenic'] > 0.3, 1, 0)
hm_df['mrl_mercury'] = np.where(hm_df['Mercury'] > 1, 1, 0)
hm_df['mrl_copper'] = np.where(hm_df['Copper'] > 30, 1, 0)
hm_df.head()

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  mrl_lead  mrl_cadmium  mrl_arsenic  mrl_mercury  \
0     0.032    40     0.0         0            0            0            0   
1     0.029    33     0.0         0            0            0            0   
2     0.038    37     0.0         0            0            0            0   
3     0.031    37     0.0         0            0            0            0   
4     0.052    35     0.0         0            0            0            0   

   mrl_copper  
0           0  
1           0  
2           0  
3           0  
4           0

### Pesticides

Fungicide :

In [11]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'fungicide'].head(5)

importName                       name    statusEN   LMR  \
2         Ametoctradin               ametoctradin  authorized  0.05   
4         Azoxystrobin               azoxystrobin  authorized  0.05   
5            Benalaxyl                  benalaxyl  prohibited  0.05   
7  Benthiavalicarb.iso  benthiavalicarb-isopropyl  authorized  0.05   
8     Benzovindiflupyr           benzovindiflupyr  authorized  0.05   

             familyEN     typeEN  
2  triazolopyrimidine  fungicide  
4         strobilurin  fungicide  
5        phenylamides  fungicide  
7           carbamate  fungicide  
8               amide  fungicide

In [12]:
fungicides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'fungicide'].importName.to_list()
fungicides.append('Diphenylamine')
fungicides.append('Tebuconazole')
fungicides

['Ametoctradin',
 'Azoxystrobin',
 'Benalaxyl',
 'Benthiavalicarb.iso',
 'Benzovindiflupyr',
 'Boscalid',
 'Bupirimate',
 'Captan',
 'Carbendazim',
 'Chlorothalonil',
 'Cyazofamid',
 'Cyflufenamid',
 'Cymoxanil',
 'Cyprodinil',
 'Difenoconazole',
 'Dimethomorph',
 'Dimoxystrobin',
 'Dodine',
 'Epoxiconazole',
 'Ethirimol',
 'Famoxadone',
 'Fenamidone',
 'Fenbuconazole',
 'Fenhexamid',
 'Fenpropidin',
 'Fenpropimorph',
 'Fluazinam',
 'Fludioxonil',
 'Fluopicolide',
 'Fluopyram',
 'Flutolanil',
 'Fluxapyroxad',
 'Hexaconazole',
 'Iprovalicarb',
 'Isofetamid',
 'Mandipropamid',
 'Metalaxyl',
 'Metconazole',
 'Metrafenone',
 'Penconazole',
 'Pencycuron',
 'Pentachloroanisol',
 'Prochloraz',
 'Propamocarb',
 'Propiconazole',
 'Prothioconazole_d',
 'Pyraclostrobin',
 'Pyrimethanil',
 'Quinoxyfen',
 'Spiroxamine',
 'Tetraconazole',
 'Thiabendazole',
 'Thiophanate.methyl',
 'Trifloxystrobin',
 'X2.phenylphenol',
 'Zoxamide',
 'Diphenylamine',
 'Tebuconazole']

Insecticide :

In [13]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'insecticide'].head(5)

importName                 name          statusEN   LMR  \
1                Aldrin               aldrin        prohibited  0.01   
13  Chlorantraniliprole  chlorantraniliprole        authorized  0.05   
17  Chlorpyrifos.methyl  chlorpyrifos-methyl        prohibited  0.05   
19            Coumaphos            coumaphos        prohibited  0.10   
20         Cyanofenphos         cyanofenphos  not yet assessed   NaN   

               familyEN       typeEN  
1        organochlorine  insecticide  
13  anthranilic diamide  insecticide  
17      organophosphate  insecticide  
19        thiophosphate  insecticide  
20      organophosphate  insecticide

In [14]:
insecticides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'insecticide'].importName.to_list()
insecticides.append('Diphenylamine')
insecticides.append('Methiocarb_total')
insecticides

['Aldrin',
 'Chlorantraniliprole',
 'Chlorpyrifos.methyl',
 'Coumaphos',
 'Cyanofenphos',
 'Cyflumetofen',
 'Cypermethrin',
 'Deet',
 'Dieldrin',
 'Etofenprox',
 'Fenoxycarb',
 'Fenpropathrin',
 'Flonicamid',
 'Flucythrinate',
 'Flupyradifurone',
 'Imidacloprid',
 'Indoxacarb',
 'Methoxyfenozide',
 'Phosmet',
 'Piperonylbutoxide',
 'Pirimicarb',
 'Spinosad',
 'Spirotetramat',
 'Tau.fluvalinate',
 'Tebufenozide',
 'Tetradifon',
 'Thiacloprid',
 'Diphenylamine',
 'Methiocarb_total']

Herbicide : 

In [15]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'herbicide'].head(5)

importName           name    statusEN   LMR          familyEN  \
0       Aclonifen      aclonifen  authorized  0.05    diphenyl ether   
6     Benfluralin    benfluralin  authorized  0.05    dinitroaniline   
15  Chlorotoluron  chlorotoluron  authorized  0.05        phenylurea   
16   Chlorpropham   chlorpropham  prohibited  0.05         carbamate   
18      Clethodim      clethodim  authorized  0.05  cyclohexanedione   

       typeEN  
0   herbicide  
6   herbicide  
15  herbicide  
16  herbicide  
18  herbicide

In [16]:
herbicides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'herbicide'].importName.to_list()
herbicides

['Aclonifen',
 'Benfluralin',
 'Chlorotoluron',
 'Chlorpropham',
 'Clethodim',
 'Cycloate',
 'Diflufenican',
 'Ethofumesate',
 'Fluazifop',
 'Flufenacet',
 'Metamitron',
 'Methabenzthiazuron',
 'Metobromuron',
 'Metolachlor',
 'Nicosulfuron',
 'Pendimethalin',
 'Phenmedipham',
 'Propanil',
 'Propyzamide',
 'Prosulfocarb',
 'Prosulfuron',
 'Simazine',
 'Terbuthylazine',
 'Tri.allate',
 'Trifluralin']

Others :

In [17]:
classe_pesticides_df[(classe_pesticides_df['typeEN'] != 'fungicide') & 
                     (classe_pesticides_df['typeEN'] != 'insecticide') & 
                     (classe_pesticides_df['typeEN'] != 'herbicide')]

importName           name    statusEN   LMR  \
3           Anthraquinone  anthraquinone  prohibited  0.02   
34          Diphenylamine  diphenylamine  prohibited  0.05   
69       Methiocarb_total     methiocarb  prohibited  0.05   
86             Propargite     propargite  prohibited  0.05   
100          Tebuconazole   tebuconazole  authorized  0.05   
112  X6.benzylaminopurine  benzyladenine  authorized   NaN   
114                   NaN            NaN         NaN   NaN   

                            familyEN  \
3    polycyclic aromatic hydrocarbon   
34                             amine   
69                         carbamate   
86                               NaN   
100                         triazole   
112                       cytokinins   
114                              NaN   

                                           typeEN  
3                                  bird repellant  
34    growth regulator, fungicide and insecticide  
69   insecticide, molluscicide and bird repellent  
86                                      acaricide  
100                fungicide and growth regulator  
112                              growth regulator  
114                                           NaN

##### Choice to treat the others

Adding to the corresponding categories, stay in "other" ? What about the growth regulator ?

#### Add some columns

In [18]:
pesticides_df['qtt_fungicides'] = pesticides_df.apply(qtt_pesticides, args = fungicides, axis = 1)
pesticides_df['qtt_insecticides'] = pesticides_df.apply(qtt_pesticides, args = insecticides, axis = 1)
pesticides_df['qtt_herbicides'] = pesticides_df.apply(qtt_pesticides, args = herbicides, axis = 1)

In [19]:
pesticides_df['nb_fungicides'] = pesticides_df.apply(nb_pesticides, args = fungicides, axis = 1)
pesticides_df['nb_insecticides'] = pesticides_df.apply(nb_pesticides, args = insecticides, axis = 1)
pesticides_df['nb_herbicides'] = pesticides_df.apply(nb_pesticides, args = herbicides, axis = 1)

In [20]:
pesticides_df['lmr_fungicides'] = pesticides_df.apply(lmr_pesticides, args = fungicides, axis = 1)
pesticides_df['lmr_insecticides'] = pesticides_df.apply(lmr_pesticides, args = insecticides, axis = 1)
pesticides_df['lmr_herbicides'] = pesticides_df.apply(lmr_pesticides, args = herbicides, axis = 1)

In [21]:
pesticides_df.head(10)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
0            B0017  2017/P1     0.0     0.0000          0.00           0.0   
1            B0019  2017/P1     0.0     0.0188          0.00           0.0   
2            B0021  2017/P1     0.0     0.0711          0.00           0.0   
3            B0023  2017/P1     0.0     0.0294          0.00           0.0   
4            B0037  2017/P1     0.0     0.0000          0.01           0.0   
5            B0039  2017/P1     0.0     0.0000          0.00           0.0   
6            B0041  2017/P1     0.0     0.0000          0.00           0.0   
7            B0043  2017/P1     0.0     0.0000          0.00           0.0   
8            B0045  2017/P1     0.0     0.0000          0.00           0.0   
9            B0059  2017/P1     0.0     0.0000          0.00           0.0   

   Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  Pentachloroanisol  \
0          0.00           0.0         0.00        0.0  ...                0.0   
1          0.00           0.0         0.00        0.0  ...                0.0   
2          0.00           0.0         0.00        0.0  ...                0.0   
3          0.00           0.0         0.00        0.0  ...                0.0   
4          0.00           0.0         0.00        0.0  ...                0.0   
5          0.01           0.0         0.00        0.0  ...                0.0   
6          0.00           0.0         0.00        0.0  ...                0.0   
7          0.00           0.0         0.01        0.0  ...                0.0   
8          0.00           0.0         0.00        0.0  ...                0.0   
9          0.00           0.0         0.00        0.0  ...                0.0   

   qtt_fungicides  qtt_insecticides  qtt_herbicides  nb_fungicides  \
0          0.0000            0.0000            0.00              0   
1          0.0188            0.0000            0.00              1   
2          0.0711            0.0000            0.00              1   
3          0.0294            0.0000            0.00              1   
4          0.0000            0.0000            0.01              0   
5          0.0000            0.0208            0.00              0   
6          0.0000            0.0000            0.00              0   
7          0.0100            0.0000            0.00              1   
8          0.0000            0.0000            0.00              0   
9          0.0145            0.0000            0.00              1   

   nb_insecticides  nb_herbicides  lmr_fungicides  lmr_insecticides  \
0                0              0               0                 0   
1                0              0               0                 0   
2                0              0               1                 0   
3                0              0               0                 0   
4                0              1               0                 0   
5                2              0               0                 0   
6                0              0               0                 0   
7                0              0               0                 0   
8                0              0               0                 0   
9                0              0               0                 0   

   lmr_herbicides  
0               0  
1               0  
2               0  
3               0  
4               0  
5               0  
6               0  
7               0  
8               0  
9               0  

[10 rows x 126 columns]

In [22]:
pesticides_df['year'] = pesticides_df['PERIOD'].apply(lambda x: x[:4])
pesticides_df['trimester'] = pesticides_df['PERIOD'].apply(lambda x: x[-1:])
pesticides_df.sample(10)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
285            B0043  2019/P3     0.0     0.0000           0.0           0.0   
417            F0019  2020/P2     0.0     0.0000           0.0           0.0   
329            F0025  2019/P4     0.0     0.0107           0.0           0.0   
16             F0005  2017/P1     0.0     0.0221           0.0           0.0   
92             B0105  2018/P1     0.0     0.0000           0.0           0.0   
174            F0029  2018/P3     0.0     0.0000           0.0           0.0   
81             B0037  2018/P1     0.0     0.0000           0.0           0.0   
197            B0021  2019/P1     0.0     0.0000           0.0           0.0   
40             F0007  2017/P2     0.0     0.0000           0.0           0.0   
409            F0001  2020/P2     0.0     0.0000           0.0           0.0   

     Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  \
285           0.0           0.0       0.0000     0.0000  ...   
417           0.0           0.0       0.0000     0.0000  ...   
329           0.0           0.0       0.0000     0.0000  ...   
16            0.0           0.0       0.0000     0.0000  ...   
92            0.0           0.0       0.0000     0.0000  ...   
174           0.0           0.0       0.0000     0.0000  ...   
81            0.0           0.0       0.0000     0.0000  ...   
197           0.0           0.0       0.0000     0.0000  ...   
40            0.0           0.0       0.0000     0.0000  ...   
409           0.0           0.0       0.0062     0.0732  ...   

     qtt_insecticides  qtt_herbicides  nb_fungicides  nb_insecticides  \
285            0.0000             0.0              2                0   
417            0.0000             0.0              2                0   
329            0.0000             0.0              1                0   
16             0.0100             0.0              1                1   
92             0.0000             0.0              0                0   
174            0.0000             0.0              0                0   
81             0.0000             0.0              2                0   
197            0.0000             0.0              0                0   
40             0.0000             0.0              0                0   
409            1.4054             0.0             14                2   

     nb_herbicides  lmr_fungicides  lmr_insecticides  lmr_herbicides  year  \
285              0               0                 0               0  2019   
417              0               1                 0               0  2020   
329              0               0                 0               0  2019   
16               0               0                 0               0  2017   
92               0               0                 0               0  2018   
174              0               0                 0               0  2018   
81               0               1                 0               0  2018   
197              0               0                 0               0  2019   
40               0               0                 0               0  2017   
409              0               1                 1               0  2020   

     trimester  
285          3  
417          2  
329          4  
16           1  
92           1  
174          3  
81           1  
197          1  
40           2  
409          2  

[10 rows x 128 columns]

##### Alternate (smaller) dataset for pesticides

In [23]:
pesti_names = pesticides_df.columns.to_list()[1:-11]

In [24]:
small_pesticides_df = pesticides_df.drop(pesti_names, axis=1)
small_pesticides_df.head(5)

REF....SUBSTANCE  qtt_fungicides  qtt_insecticides  qtt_herbicides  \
0            B0017          0.0000               0.0            0.00   
1            B0019          0.0188               0.0            0.00   
2            B0021          0.0711               0.0            0.00   
3            B0023          0.0294               0.0            0.00   
4            B0037          0.0000               0.0            0.01   

   nb_fungicides  nb_insecticides  nb_herbicides  lmr_fungicides  \
0              0                0              0               0   
1              1                0              0               0   
2              1                0              0               1   
3              1                0              0               0   
4              0                0              1               0   

   lmr_insecticides  lmr_herbicides  year trimester  
0                 0               0  2017         1  
1                 0               0  2017         1  
2                 0               0  2017         1  
3                 0               0  2017         1  
4                 0               0  2017         1

##### Alternate (group by) dataset for pesticides

In [25]:
group_pesti_df = pesticides_df.groupby('REF....SUBSTANCE').agg(max)

In [26]:
group_pesti_df = group_pesti_df.loc[:,'lmr_fungicides' : 'lmr_herbicides']
group_pesti_df

lmr_fungicides  lmr_insecticides  lmr_herbicides
REF....SUBSTANCE                                                  
B0012                          1                 0               0
B0017                          1                 0               0
B0019                          0                 0               0
B0021                          1                 0               0
B0023                          0                 0               0
B0032                          1                 0               0
B0037                          1                 1               0
B0039                          1                 0               0
B0041                          1                 0               0
B0043                          1                 0               1
B0045                          1                 0               0
B0059                          1                 0               0
B0061                          0                 0               0
B0067                          1                 0               0
B0069                          1                 1               0
B0077                          1                 1               0
B0093                          1                 0               0
B0096                          0                 0               0
B0103                          1                 0               0
B0105                          1                 1               1
B0116                          0                 0               0
B0128                          0                 0               0
B0140                          0                 0               0
B0142                          1                 0               0
B0152                          1                 0               0
B0158                          1                 0               0
B0163                          1                 0               0
F0001                          1                 1               0
F0003                          1                 0               0
F0005                          0                 0               0
F0007                          0                 0               0
F0009                          0                 0               0
F0011                          0                 1               0
F0013                          0                 0               0
F0015                          1                 0               0
F0017                          0                 0               0
F0019                          1                 0               0
F0020                          1                 0               0
F0021                          1                 0               0
F0023                          0                 0               1
F0025                          0                 0               0
F0029                          1                 0               0
F0031                          1                 1               0
F0033                          0                 0               0
F0035                          0                 0               1
F0039                          1                 0               0
F0041                          0                 0               0
F0043                          0                 0               0
F0055                          1                 0               0
F0057                          1                 0               0
F0061                          1                 0               0
F0063                          1                 0               0
F0065                          0                 0               0
F0069                          0                 0               1
F0071                          1                 0               0
F0073                          1                 1               0
FE0001                         0                 0               0

### Sites

##### Getting the area for each zone

In [27]:
distances_df.head(5)

Site     polyID  classCLC         dist
0  F0069   FR-88090       231  1183.022362
1  F0069  FR-131773       242   289.926324
2  F0069   FR-46167       211   221.344153
3  F0069   FR-11047       112  1038.383069
4  F0069  FR-195343       311   724.045683

In [28]:
surfaces_df.head(5)

Site  111           112           121            122  123  124  131  132  \
0  F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1  F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2  F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3  B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4  F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   

   133  ...  411  412  421  422  423  511            512  521  522  523  
0  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4  0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  

[5 rows x 45 columns]

In [29]:
# surfaces_df['country'] = surfaces_df['Site'].apply(lambda x: 'FR' if x.startswith('F') else 'BE')
# surfaces_df['country'].value_counts()

In [30]:
sites_df = distances_df

In [31]:
sites_df['surface'] = sites_df.apply(get_surface, axis = 1)

In [32]:
sites_df.sample(10)

Site     polyID  classCLC         dist       surface
360  F0033  FR-239285       313   351.169291  2.026127e+06
208  B0045   BE-24952       523  1217.224329  2.535093e+05
54   F0071  FR-176936       311  1074.120276  5.607325e+05
569  F0009  FR-158678       243  1381.044936  2.040207e+04
509  F0063  FR-206750       311    82.223549  1.751218e+06
227  B0037   BE-18145       242  1053.089712  1.335673e+06
575  F0011  FR-122016       242   850.371468  1.525860e+06
59   F0071  FR-254625       324  1090.453025  7.933020e+05
185  B0140      BE-24       111   408.916661  1.011569e+06
535  F0013  FR-158678       243   420.039493  2.507248e+05

In [33]:
sites_df['country'] = sites_df['polyID'].apply(lambda x: 'FR' if x.startswith('FR') else 'BE')
sites_df.sample(10)

Site     polyID  classCLC         dist       surface country
581  F0011  FR-187751       311    24.499390  2.443109e+06      FR
154  F0075   FR-29212       121   344.517523  4.422800e+05      FR
271  B0059   BE-11244       231   622.842459  2.062409e+06      BE
209  B0041      BE-48       111   701.369556  2.742902e+05      BE
443  B0103    BE-4734       131     0.000000  1.160708e+06      BE
338  F0029  FR-147692       242  1351.573196  7.740918e+05      FR
153  F0075   FR-20030       112   217.181706  1.687705e+06      FR
320  F0017   FR-48582       211   689.236741  9.580098e+05      FR
314  F0017   FR-90989       231    74.643842  4.409383e+06      FR
610  B0152    BE-6984       211   649.945771  1.121794e+06      BE

In [34]:
sites_df['country'].value_counts()

FR    335
BE    297
Name: country, dtype: int64

In [35]:
sites_df = sites_df.drop(['polyID'], axis=1)
sites_df.sample(5)

Site  classCLC         dist       surface country
617  F0003       313   228.681874  1.955361e+06      FR
610  B0152       211   649.945771  1.121794e+06      BE
300  F0055       112   892.753017  1.290453e+06      FR
161  F0065       141  1202.898948  2.575788e+05      FR
548  F0015       312    80.820080  1.651150e+06      FR

##### Get one dataset with : beehives and area/distance for each zone

In [36]:
surfaces_df

Site  111           112           121            122  123  124  131  132  \
0   F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1   F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2   F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3   B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4   F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   
..    ...  ...           ...           ...            ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00       0.000000  0.0  0.0  0.0    0   
59  B0069  0.0  8.704243e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
60  B0152  0.0  4.575279e+06  5.726391e+05  201434.000214  0.0  0.0  0.0    0   
61  F0003  0.0  3.425116e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
62  B0158  0.0  1.467398e+05  0.000000e+00  318178.187439  0.0  0.0  0.0    0   

    133  ...  411  412  421  422  423  511            512  521  522  523  
0   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4   0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...            ...  ...  ...  ...  
58  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
59  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
60  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
61  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
62  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  

[63 rows x 45 columns]

In [37]:
classCLC = surfaces_df.columns.to_list()[1:]

In [38]:
by_sites_df = surfaces_df.copy(deep=True)
by_sites_df

Site  111           112           121            122  123  124  131  132  \
0   F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1   F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2   F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3   B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4   F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   
..    ...  ...           ...           ...            ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00       0.000000  0.0  0.0  0.0    0   
59  B0069  0.0  8.704243e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
60  B0152  0.0  4.575279e+06  5.726391e+05  201434.000214  0.0  0.0  0.0    0   
61  F0003  0.0  3.425116e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
62  B0158  0.0  1.467398e+05  0.000000e+00  318178.187439  0.0  0.0  0.0    0   

    133  ...  411  412  421  422  423  511            512  521  522  523  
0   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4   0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...            ...  ...  ...  ...  
58  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
59  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
60  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
61  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
62  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  

[63 rows x 45 columns]

In [39]:
for CLC in classCLC:
    by_sites_df[CLC] = by_sites_df.apply(distance_area, args = (CLC,), axis = 1)

In [40]:
by_sites_df

Site  111           112           121         122  123  124  131  132  \
0   F0069  0.0  1.442780e+03  3.587087e+05    0.000000  0.0  0.0  0.0  0.0   
1   F0025  0.0  0.000000e+00  0.000000e+00  700.801763  0.0  0.0  0.0  0.0   
2   F0027  0.0  2.604161e+02  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
3   B0128  0.0  2.412962e+03  4.053755e+06  523.436355  0.0  0.0  0.0  0.0   
4   F0023  0.0  0.000000e+00  7.907522e+02    0.000000  0.0  0.0  0.0  0.0   
..    ...  ...           ...           ...         ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
59  B0069  0.0  7.328884e+03  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
60  B0152  0.0  4.575279e+06  5.227880e+02  421.402885  0.0  0.0  0.0  0.0   
61  F0003  0.0  3.859675e+02  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
62  B0158  0.0  1.574508e+02  0.000000e+00  392.706036  0.0  0.0  0.0  0.0   

    133  ...  411  412  421  422  423  511         512  521  522  523  
0   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
1   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
2   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
3   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
4   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  463.344615  0.0  0.0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...         ...  ...  ...  ...  
58  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
59  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
60  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
61  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
62  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  

[63 rows x 45 columns]

check if correct

In [41]:
distance = distances_df[(distances_df['Site'] == 'F0023') & 
                        (distances_df['classCLC'] == 512)]['dist'].values
distance

array([614.50674791])

In [42]:
area = surfaces_df[surfaces_df['Site'] == 'F0023']['512'].values[0]
area

284728.392578125

In [43]:
area/distance[0]

463.3446150839585

### Merging the datasets into two

3 datasets to 2 : sites + (heavy metal || pesticides)

##### Heavy Metal

##### Pesticides

full

In [44]:
pesticides_df.sample(5)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
451            B0103  2020/P3     0.0      0.031           0.0           0.0   
14             F0001  2017/P1     0.0      0.000           0.0           0.0   
291            B0096  2019/P3     0.0      0.000           0.0           0.0   
36             B0105  2017/P2     0.0      0.000           0.0           0.0   
53             B0043  2017/P3     0.0      0.000           0.0           0.0   

     Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  \
451           0.0           0.0          0.0        0.0  ...   
14            0.0           0.0          0.0        0.0  ...   
291           0.0           0.0          0.0        0.0  ...   
36            0.0           0.0          0.0        0.0  ...   
53            0.0           0.0          0.0        0.0  ...   

     qtt_insecticides  qtt_herbicides  nb_fungicides  nb_insecticides  \
451               0.0             0.0              4                0   
14                0.0             0.0              0                0   
291               0.0             0.0              1                0   
36                0.0             0.0              0                0   
53                0.0             0.0              0                0   

     nb_herbicides  lmr_fungicides  lmr_insecticides  lmr_herbicides  year  \
451              0               1                 0               0  2020   
14               0               0                 0               0  2017   
291              0               0                 0               0  2019   
36               0               0                 0               0  2017   
53               0               0                 0               0  2017   

     trimester  
451          3  
14           1  
291          3  
36           2  
53           3  

[5 rows x 128 columns]

In [45]:
small_pesticides_df.sample(5)

REF....SUBSTANCE  qtt_fungicides  qtt_insecticides  qtt_herbicides  \
514            F0057          0.0000            0.0000          0.0204   
433            F0073          6.3265            0.4300          0.0287   
284            B0041          0.0000            0.0044          0.0000   
133            F0017          0.0000            0.0000          0.0000   
392            B0039          0.0428            0.0000          0.0058   

     nb_fungicides  nb_insecticides  nb_herbicides  lmr_fungicides  \
514              0                0              1               0   
433              4                1              1               1   
284              0                1              0               0   
133              0                0              0               0   
392              4                0              1               0   

     lmr_insecticides  lmr_herbicides  year trimester  
514                 0               0  2020         4  
433                 1               0  2020         2  
284                 0               0  2019         3  
133                 0               0  2018         2  
392                 0               0  2020         2

In [46]:
group_pesti_df.sample(5)

lmr_fungicides  lmr_insecticides  lmr_herbicides
REF....SUBSTANCE                                                  
F0011                          0                 1               0
B0017                          1                 0               0
F0031                          1                 1               0
F0055                          1                 0               0
F0035                          0                 0               1

##### sites

In [47]:
sites_df.sample(5)

Site  classCLC         dist       surface country
451  B0103       231  1484.694077  4.801663e+05      BE
505  F0061       211   303.450850  1.270137e+06      FR
603  B0069       311  1499.087943  2.383795e+05      BE
628  B0158       231  1103.848468  4.010538e+05      BE
340  F0029       131   392.579026  2.951948e+05      FR

In [48]:
by_sites_df.sample(5)

Site         111           112          121          122  123  124  131  \
57  F0011    0.000000  7.531979e+02     0.000000     0.000000  0.0  0.0  0.0   
23  B0041  391.077976  2.904582e+06   154.142379     0.000000  0.0  0.0  0.0   
27  F0035    0.000000  0.000000e+00     0.000000     0.000000  0.0  0.0  0.0   
9   B0142    0.000000  6.808838e+03  6009.480699  1078.987269  0.0  0.0  0.0   
49  F0061    0.000000  4.997552e+03    82.566983     0.000000  0.0  0.0  0.0   

    132  133  ...  411  412  421  422  423  511  512  521  522  523  
57  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
23  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
27  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
9   0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
49  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 45 columns]

#### Merge

### Changing the NA

# Questions to ask

- When NaN => information irrelevant or not analysed?
- Difference France/Belgium?
- Is the fact that an insecticide is unauthorized relevant? (=> not the same law in France/Belgium)
- Or do we have to show that even banned pesticides are still found?
- How do we treat other categorical pesticides?
- How to increase the number of datas?

### Bo-bo-bonus

- FR  :  36 beehives for 335 zones
- BE  :  27 beehives for 297 zones

Here the CLC zone are left separate, could group some to test